### DISCUSSION TITLE
scripted treatment-plan optimization - definition of objectives/constraints

### DISCUSSION THREAD

**Discussion by ehssal** [e0404/matRad#517] [Q&A]
Hello,

I am new to matRad and am looking for a way to define the treatment-plan optimization problem within the MATLAB scripting environment (as opposed to matRadGUI). How can I select and define the constraints as well as the objective function terms using the appropriate functions? Is there an example available?

**Comment by wahln** [comment]
Hi,
in the latest version the objectives/constraints are implemented as classes in the optimization/+DoseObjectives & optimization/+DoseConstraints folders (the + indicates a Matlab package folder which acts as namespace).
For the optimization, these need to be put in the 6-th column of the `cst` in the row of the respective Segmentation, where all objectives are stored in a cell array. A squared deviation objective with reference dose 60 Gy & penalty 1000 in the first structure, for example, could be created by 
`cst{1,6}{1} = DoseObjectives.matRad_SquaredDeviation(1000,60)`
This should work with modern Matlab versions.
A more safe way to be compatible with different mat-file versions and so on, is to turn the objectives into structs, which avoids them being saved as Matlab objects:
`cst{1,6}{1} = struct(DoseObjectives.matRad_SquaredDeviation(1000,60))`
This is the way the standard objectives are stored in the given patient files, but both ways work.

There should be some similar code in the first example. and a little explanation in the wiki: https://github.com/e0404/matRad/wiki/Fluence-optimization
